docs/development/design/telemetry.rst

```diff
@@ -0,0 +1,232 @@
+   base-passwd                                     install
+   bash                                            install
+   binutils                                        install
+   binutils-common:amd64                           install
+   binutils-x86-64-linux-gnu                       install
+   bsdutils                                        install
+   build-essential                                 install
+
+Python
+~~~~~~
+
+We can get the Python version from the config file when using a Python environment,
+and from the ``conda list`` output when using a Conda environment.
+
+OS
+~~
+
+We can infer the OS version from the build image used in the config file,
+but since it changes with time, we can get it from the OS itself:
+
+.. code-block::
+
+   $ lsb_release --description
+   Description:    Ubuntu 18.04.5 LTS
+   # or
+   $ cat /etc/issue
+   Ubuntu 18.04.5 LTS \n \l
+
+Storage
+-------
+
+We can save all this information in json files in cloud storage,
+then we could use a tool to import all this data into.
+Or we can decide for a tool or service where to fed all this data directly into.
+
+If we decide to save the files in cloud storage,
+we can try to calculate a hash of the file so we don't upload duplicates that happen on the same day/month.
+We can aggregate this data per year/month saving them in following structure:
+``telemetry/builds/{year}/{month}/{year}-{month}-{day}-{timestamp-pk|pk}.json``,
+that way is easy to download, all data per year/month without iterating over all files.
+
+.. Since this information isn't sensitive,
+   I think we are fine with this structure
+   (we can't do bulk deletes of all info about a project if we follow this structure).
+
+Format
+~~~~~~
+
+The final file to be saved would have the following information:
+
+- project: the project slug
+- version: the version slug
+- build: the build id (which may stop existing if the project is deleted)
+- date: full date in isoformat or timestamp (POSIX)
+- user_config: Original user config file
+- final_config: Final configuration used (merged with defaults)
+- packages.pip: List of pip packages with name and version
+- packages.conda: List of conda packages with name, channel, and version
+- packages.apt: List of apt packages
+- python: Python version used
+- os: Operating system used
+
+.. code-block:: json
+
+   {
+     "project": "docs",
+     "version": "latest",
+     "build": 12,
+     "date": "2021-04-20-...",
+     "user_config": {},
+     "final_config": {},
+     "packages": {
+        "pip": [{
+           "name": "sphinx",
+           "version": "3.4.5"
+        }],
+        "conda": [{
+           "name": "sphinx",
+           "channel": "conda-forge",
+           "version": "0.1"
+        }],
+        "apt": [
+            "python3-dev",
+            "cmatrix"
+        ]
+     },
+     "python": "3.7",
+     "os": {
+         "name": "ubuntu",
+         "version": "18.04.5"
+     }
+   }
+
+Analyzing the data
+------------------
+
```

Comment:
I think we actually probably need to have this section first. Knowing what kind of analysis we want to run on the data will tell us how to store & query it. 

Currently we have defined a way of storing data that we don't really look at. I think it's probably more valuable to start to have some kind of short-term storage of data in a query engine that we can query dynamically (ES or Postgres, as @astrojuanlu said). We'd have to pretty strongly truncate this data (eg. <4GB in memory) -- but I think we can definitely get some useful subsets of data into that format given that it isn't huge.

In particular, just getting this data into some kind of queryable format:

* .readthedocs.yml
* conf.py dumped as JSON
* pip & conda files dumped

Would get us a lot of value across a short timeframe (eg. 1 month).

But, what are we going to do when we have this data to query? We should think about what kind of queries we want to run and what analysis would be valuable to make sure we're building the system properly.